### ISSUE TITLE
MappedAs throws when called on a Convert UnaryExpression

### ISSUE THREAD

**Issue by alexrootes** [nhibernate/nhibernate-core#2708]
In an apparent regression between `5.2.7` and `5.3.0` this code will throw where it did not before:

```c#
q = q.Where(r => r.Labels == ((IntArray)labels).MappedAs(CustomType.IntArrayType));
```
```
HibernateException: MappedAs must be called on an expression which can be evaluated as ConstantExpression. It was call on UnaryExpression instead.
```
As a workaround removing the inline cast resolves the issue:
```c#
var convertedLabels = (IntArray)labels;
q = q.Where(r => r.Labels == convertedLabels.MappedAs(CustomType.IntArrayType));
```
This could possibly be a result of changes in #2365

**Comment by fredericDelaporte** [comment]
Have you checked if the trouble does still occur with 5.3.7?

**Comment by alexrootes** [comment]
Yes same issue under 5.3.7.

**Comment by bahusoid** [comment]
Issue aside. Is `MappedAs` still required for this case in 5.3? I believe #2365 should make it work without explicit `MappedAs` call.

**Comment by alexrootes** [comment]
Thanks @bahusoid I can confirm that code does indeed work. I did simplify the original code to better illustrate the issue however, the full version looks like this:
```c#
q.Where(r => r.Labels.Contains(((IntArray)labels).MappedAs(CustomType.IntArrayType)));
```
Where `Contains` is a custom Linq method hooked to an HQL function generator. Omitting the `MappedAs` call from that throws:
```
HibernateException: Could not determine a type for class: IntArray
```
I would like to understand how to resolve that but it's probably outside the scope of this issue.

I can workaround by directly constructing an IntArray instance rather than converting:
```c#
q.Where(r => r.Labels.Contains(new IntArray(labels).MappedAs(CustomType.IntArrayType)));
```

**Comment by bahusoid** [comment]
Hm... Not sure how it supposed to work for custom generators. Also your original code does not look right - `Contains` expect collection element as a parameter - from your example it seems you supply collections on both sides of `Contains`.

From quick glance at code it seems auto detection is currently implemented only for inversed case:
```C#
q.Where( r = > new IntArray(labels).Contains(r.Label)); //Contains is called on parameter, not on mapped property
````
So `MappedAs` might be still required in your case.

**Comment by alexrootes** [comment]
Yes this is why `Contains` is a custom Linq method, it boils down to a pg array contains operator `@>`  https://www.postgresql.org/docs/current/functions-array.html#ARRAY-OPERATORS-TABLE

**Comment by bahusoid** [comment]
> I would like to understand how to resolve that but it's probably outside the scope of this issue

You can map custom type using `TypeFactory.RegisterType`. Something like:
```C#
//Before building NHibernate configuration
TypeFactory.RegisterType(typeof(IntArray), CustomType.IntArrayType, TypeFactory.EmptyAliases);
```

**Comment by fredericDelaporte** [comment]
Fixed by #2711
